### PR TITLE
feat(GODT-2289): Use timestamp for UIDValidity

### PR DESCRIPTION
**Pull request by LBeernaertProton**
Use a epoch timestamp for uidvalidity starting from 1st of February
2023. This will ensure that the user will always have unique UIDValidities even if the connector data is erased/lost.

### CODE REVIEW

File: imap/uid_validity_generator.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func DefaultEpochUIDValidityGenerator() *EpochUIDValidityGenerator {
+	return NewEpochUIDValidityGenerator(time.Date(2023, 2, 1, 0, 0, 0, 0, time.UTC))
+}
+
+func (e *EpochUIDValidityGenerator) Generate() (UID, error) {
+	timeStamp := uint64(time.Now().Sub(e.epochStart).Seconds())
+	if timeStamp > uint64(0xFFFFFFFF) {
+		return 0, fmt.Errorf("failed to generate uid validaty, interval exceeded maximum capacity")
```
Comment (jameshoulahan) [inline imap/uid_validity_generator.go:31]:
```suggestion
		return 0, fmt.Errorf("failed to generate UID validity, interval exceeds maximum size")
```

File: imap/uid_validity_generator.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	// This loops is here to ensure that two successive calls to Generate that happen during the same second
+	// can still generate unique values. To avoid waiting another second until the values are different,
+	// we keep bumping the last generated value until it is greater than the last generated value.
+	for {
+		lastGenerated := atomic.LoadUint32(&e.lastUID)
+
+		// Not enough time elapsed between the last time
+		if lastGenerated >= timeStampU32 {
+			if timeStampU32 == 0xFFFFFFFF {
+				return 0, fmt.Errorf("failed to generate uid validaty, interval exceeded maximum capacity")
```
Comment (jameshoulahan) [inline imap/uid_validity_generator.go:45]: Same error, i would define `var ErrUIDValidityTooLarge = errors.New("failed to generate UID validity, interval exceeds maximum size")`

File: internal/backend/state_user_interface_impl.go [hunk 3]
```diff
@@ -2,8 +2,8 @@ package backend
 
 import (
 	"context"
-
 	"github.com/ProtonMail/gluon/imap"
+
```
Comment (jameshoulahan) [inline internal/backend/state_user_interface_impl.go:6]: no need for this line

File: internal/state/user_interface.go [hunk 4]
```diff
@@ -2,8 +2,8 @@ package state
 
 import (
 	"context"
-
 	"github.com/ProtonMail/gluon/imap"
+
```
Comment (jameshoulahan) [inline internal/state/user_interface.go:6]: no need for this line

File: option.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package gluon
 
 import (
 	"crypto/tls"
+	"github.com/ProtonMail/gluon/imap"
```
Comment (jameshoulahan) [inline option.go:5]: More weird header ordering 🤔 maybe your editor is using a different autoformatter than mine. Gluon stuff should be below the stdlib stuff.

File: server.go [hunk 6]
```diff
@@ -6,6 +6,7 @@ import (
 	"crypto/tls"
 	"errors"
 	"fmt"
+	"github.com/ProtonMail/gluon/imap"
```
Comment (jameshoulahan) [inline server.go:9]: Again, import grouping

File: internal/state/user_interface.go (hunk 4 above)
Comment (LBeernaertProton) [reply internal/state/user_interface.go]: Done

File: internal/backend/state_user_interface_impl.go (hunk 3 above)
Comment (LBeernaertProton) [reply internal/backend/state_user_interface_impl.go]: Done

File: imap/uid_validity_generator.go (hunk 2 above)
Comment (LBeernaertProton) [reply imap/uid_validity_generator.go]: Done

File: imap/uid_validity_generator.go (hunk 1 above)
Comment (LBeernaertProton) [reply imap/uid_validity_generator.go]: Done

File: server.go (hunk 6 above)
Comment (LBeernaertProton) [reply server.go:9]: Done

File: connector/dummy_simulate.go [hunk 7]
```diff
@@ -217,8 +217,7 @@ func (conn *Dummy) MessageDeleted(messageID imap.MessageID) error {
 }
 
 func (conn *Dummy) UIDValidityBumped() {
```
Comment (cuthix) [inline connector/dummy_simulate.go:219]: Do we still need this?

Comment (LBeernaertProton) [reply connector/dummy_simulate.go:219]: Yes. it is still used to test UID validity bumped or if we need to force GLuon to regenerate for whatever reason.

**Review by jameshoulahan — APPROVED** [review]

**Review by cuthix — APPROVED** [review]

**Pull request merged**